compiler/luci/pass/src/FuseHorizontalFCLayers.cpp

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2023 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FuseHorizontalFCLayers.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Service/Nodes/CircleConst.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+namespace
+{
+bool check_type_and_shape_equality(luci::CircleNode *left, luci::CircleNode *right)
+{
+  if (left->dtype() != right->dtype())
+    return false;
+
+  if (left->rank() != right->rank())
+    return false;
+
+  for (uint32_t i = 0; i < left->rank(); ++i)
+  {
+    if (left->dim(i).value() != right->dim(i).value())
+      return false;
+  }
+
+  return true;
+}
+
+template <loco::DataType D> void sum_const_values(luci::CircleConst *left, luci::CircleConst *right)
+{
+  const auto size = left->template size<D>();
+
+  for (uint32_t i = 0; i < size; ++i)
+  {
+    left->template at<D>(i) += right->template at<D>(i);
+  }
+}
+
+bool fuse_horizontal_fc_nodes(luci::CircleAdd *add_node)
+{
+  // Let's check left and right FC nodes
+  auto left_fc_node = dynamic_cast<luci::CircleFullyConnected *>(add_node->x());
+  auto right_fc_node = dynamic_cast<luci::CircleFullyConnected *>(add_node->y());
+
+  if (left_fc_node == nullptr or right_fc_node == nullptr)
+    return false;
+
+  if (not check_type_and_shape_equality(left_fc_node, right_fc_node))
+    return false;
+
+  if (left_fc_node->fusedActivationFunction() != luci::FusedActFunc::NONE)
+    return false;
+
+  if (right_fc_node->fusedActivationFunction() != luci::FusedActFunc::NONE)
+    return false;
+
+  // Let's check that FC nodes have the same input
+  auto left_input_node = dynamic_cast<luci::CircleNode *>(left_fc_node->input());
+  if (left_input_node == nullptr)
+    return false;
+
+  auto right_input_node = dynamic_cast<luci::CircleNode *>(right_fc_node->input());
+  if (right_input_node == nullptr)
+    return false;
+
+  if (left_input_node != right_input_node)
+    return false;
```

Comment:
Pointer comparison would not need casting.

```suggestion
if (left_fc_node->input() != right_fc_node->input())
  return false;
```

Reply:
Done, thank you